src/hotspot/share/gc/g1/g1FullCollector.cpp

```diff
@@ -348,37 +348,61 @@ bool G1FullCollector::phase2b_forward_oops() {
-  // to allocate the first eden region after gc.
+uint G1FullCollector::truncate_parallel_cps() {
+  uint lowest_current = (uint)-1;
+  for (uint i = 0; i < workers(); i++) {
+    G1FullGCCompactionPoint* cp = compaction_point(i);
+    if (cp->has_regions()) {
+      lowest_current = MIN2(lowest_current, cp->current_region()->hrm_index());
+    }
+  }
+
   for (uint i = 0; i < workers(); i++) {
     G1FullGCCompactionPoint* cp = compaction_point(i);
     if (cp->has_regions()) {
-      serial_compaction_point()->add(cp->remove_last());
+      cp->remove_at_or_above(lowest_current);
     }
   }
 
-  // Update the forwarding information for the regions in the serial
-  // compaction point.
-  G1FullGCCompactionPoint* cp = serial_compaction_point();
-  for (GrowableArrayIterator<HeapRegion*> it = cp->regions()->begin(); it != cp->regions()->end(); ++it) {
-    HeapRegion* current = *it;
-    if (!cp->is_initialized()) {
-      // Initialize the compaction point. Nothing more is needed for the first heap region
-      // since it is already prepared for compaction.
-      cp->initialize(current);
-    } else {
-      assert(!current->is_humongous(), "Should be no humongous regions in compaction queue");
-      G1SerialRePrepareClosure re_prepare(cp, current);
-      set_compaction_top(current, current->bottom());
-      current->apply_to_marked_objects(mark_bitmap(), &re_prepare);
+  return lowest_current;
+}
+
+void G1FullCollector::phase2c_prepare_serial_compaction() {
+  GCTraceTime(Debug, gc, phases) debug("Phase 2: Prepare serial compaction", scope()->timer());
+  // At this point, we know that after parallel compaction there will be regions that
+  // are partially compacted into. Thus, the last compaction region of all
+  // compaction queues still have space in them. We try to re-compact these regions
+  // in serial to avoid a premature OOM when the mutator wants to allocate the first
+  // eden region after gc.
+
+  // For maximum compaction, we need to re-prepare all objects above the lowest
+  // region among the current regions for all thread compaction points. It may
+  // happen that due to the uneven distribution of objects to parallel threads, holes
+  // have been created as threads compact to different target regions between the
+  // lowest and the highest region in the tails of the compaction points.
+
+  uint start_serial = truncate_parallel_cps();
+
+  G1FullGCCompactionPoint* serial_cp = serial_compaction_point();
+  HeapWord* dense_prefix_top = nullptr;
+  for (uint i = start_serial; i < _heap->max_reserved_regions(); i++) {
+    if (is_compaction_target(i)) {
+      HeapRegion* current = _heap->region_at(i);
+      serial_cp->add(current);
+      if (!serial_cp->is_initialized()) {
+        // Initialize the compaction point. Nothing more is needed for the first heap region
+        // since it is already prepared for compaction.
+        serial_cp->initialize(current);
+        dense_prefix_top = compaction_top(current);
```

Comment:
I think `serial_cp->initialize(current);` and `dense_prefix_top` can be extracted out the loop, because they correspond to `lowest_current`, which is "statically" known before entering the loop.